### DISCUSSION TITLE
Example Project : Voxel  Game

### DISCUSSION THREAD

**Discussion by ValorZard** [bevyengine/bevy#2430] [Ideas]
I think there should be a sample project for Bevy that's a bit heavier than the current example projects, while still being simple enough to build off of and make your own. Thus, I had the idea that we should make a simple voxel game, complete with tutorial.

Note: All of this is subject to change depending on community response. This is a community-led project in the first place, and I can see a lot of these systems changing as time goes on.

## Why Voxel?

There are many, many projects made in Bevy that are voxel games. In general, it's a very popular use case, and I think it would be cool to show how to generate and render them.

## Why Bevy?
Two reasons.
1. I think Bevy needs a bigger project to really test it out. There's Sotora, but that's current held up on UI improvements. You can make a shooter or a voxel game with very minimal UI.
2.  I think it would be cool to divide the code into multiple chunks. What we could do split the codebase (and the tutorial that would come along with it) into a couple of parts. 
    - The first part could be the voxel part. We show how to build and render a voxel system. This should end with something akin to an extremely stripped-down version of Minecraft's creative mode, with rolling "grass" hills.

**Comment by brianbruggeman** [comment]
IMO, any example project should be incredibly succinct.  That's why something like an early arcade game (like pong) is ideal.  FPS itself has a number of strict requirements that are orthogonal and mostly unrelated to the underlying game world structure.  I also think Voxels in particular add several layers of complexity that would detract from learning how to build a barebones FPS.  An example shouldn't be a dream game - it should just be something that can be explained well, easily expanded by a new bevy user and ultimately included within a tutorial.

I do think a very simple sandbox Voxel game would be awesome for Bevy.   And a voxel example game could borrow elements from the FPS example, but Voxel itself need not be a FPS - there are a lot of genres.

**Reply by ValorZard** [reply]
Yeah, the more i think about it the more I think the FPS thing might be unnecessary. I just put it in since it was loosely related. I'll change the topic name to more closely fit this idea.

**Reply by IceSentry** [reply]
I don't know about _any_ example project being succinct, sure most of them should be, but a slightly more advanced example with clear comments explaining what is going on is really nice when you want to make something slightly more complicated than pong.

**Reply by Nilirad** [reply]
I think that they have to be succinct in scope but still complete, with various states (main menu, pause menu, gameplay, gameover, ...).

**Reply by brianbruggeman** [reply]
I think an example of "how to build a menu" should be its own thing.  There are more than a few games that use menus, and creating a simple example that showcases menus would be ideal so that a new bevy user can pick that code up, easily understand it and then expand it.

I think a "complete" game example is really just addressing the game loop with rendering and input management.  The scope needs to be really, really small.  Extra modes should be their own example.

**Comment by Carter0** [comment]
I think a longer term project makes sense, but I think it might be more prudent to wait for a more feature complete bevy before doing this. That way you will get the benefit of building off systems that are unlikely to drastically change in the future and the game will ideally b e easier to make.

**Comment by cart** [comment]
This relates to the discussion here: https://github.com/bevyengine/bevy/issues/1248
